### ISSUE TITLE
Urls with trailing /s are not equivalent to ones without

### ISSUE THREAD

**Issue by Qrbaker** [TrueBrain/TrueWiki#108]
https://wiki.openttd.org/en and https://wiki.openttd.org/en/ (note the trailing "/") do not resolve to the same page. This breaks navigating up hierarchies.

**Comment by TrueBrain** [comment]
I am not completely sure what you mean, as you are a bit minimalistic with the ticket :) But how TrueWiki sees the world:

Everything ending on a `/` is a folder. Everything else is a file request.
A folder request, internally, is rewritten to a file request by appending `Main Page` to it (basically the `index.html` variant of what most HTTP servers do). This because we cannot render folders, only files.
(well, there are a few exceptions, mostly `/` is an exception, but okay; minor details :D).

So those two URLs, according to TrueWiki, are indeed not the same, and shouldn't be the same. `/en`, the page, doesn't exist. `/en/` internally is rewritten to `/en/Main Page` which is rendering `/Page/en/Main Page.mediawiki` from the filesystem, which does exist (most likely).

What I am not sure about, is why you wrote this ticket; what is going wrong exactly?

What do you mean with "This breaks navigating up hierarchies.". Does TrueWiki render broken breadcrumbs in the top navigation? There should never be a link to `/en` generated by TrueWiki, only to `/en/`. Did you find somewhere this not being the case?

Or is this because you have the expectation TrueWiki redirects `/en` to `/en/` automatically? In that case I wonder why this is your expectation and when one would hit `/en`?

Or is there something else behind this ticket?

All honest questions, to be clear :) Really curious about the motivation behind this question :) Tnx :)

**Comment by Qrbaker** [comment]
Ah, no. The issue is visible on our staging wiki. We made a top-level page called "Web Console" and nested some other info in `Web Console/`: 

* https://crafty-wiki.silverfyre.info/en/Web%20Console
* https://crafty-wiki.silverfyre.info/Folder/Page/en/Web%20Console/

If what you say is true, should https://crafty-wiki.silverfyre.info/en/Web%20Console/ then redirect to https://crafty-wiki.silverfyre.info/Folder/Page/en/Web%20Console/ ?

**Comment by Qrbaker** [comment]
I think I have it figured out. I moved the page to `en/Web Console/Main Page` and the hierarchy works as expected. This came down to an issue with the page creation process, which doesn't make it clear that you need to preemptively make a page name like this if you intend to nest pages under it later.

I don't know the feasibility, but it would be nice to perhaps have an option to "add a sub-page" when on the edit screen, or next to the edit button, that moves the page to a folder main page and _then_ redirects to the new page.

**Comment by TrueBrain** [comment]
Just to confirm I understand you correctly:

You expected that `en/Web Console` would be the "root" of `en/Web Console/`, so that if you visit the latter, you see the content of the first?

I can understand that; we introduced folders in a different way (as you found out by now), but I get where that idea comes from. When you hit `Create Page` on `en/Web Console/` the first time, it does add `Main Page` to it, making the suggestion how it works. But I agree you need knowledge of how folders work in order for you to understand that part.

If you would make `en/Web Console/Logs` first, clicking on the hierarchy will give you a Page Not Found for `en/Web Console/`, but hitting `Create Page` there allows you to make the right page.

But when you come from the other direction, first make `en/Web Console` and afterwards think: I want to make this a folder and have subpages, it breaks a bit.

I too need to think about this. I guess once you know, you know, but a bit of hinting how folders work wouldn't hurt anyone. Interesting :D

**Comment by Qrbaker** [comment]
Yeah, I was coming at it from a mediawiki perspective I suppose. The common case of nested pages on mediawiki sites are userpages. A user will have their `User:username` page, and might make, say, a `User:username/sandbox` page.
